### PR TITLE
added implementation and tests for crypto transfer prehandle

### LINKED ISSUE #4228: Implement `CryptoTransfer` PreTransactionHandler
We need to implement PreTransactionHandler logic that supports `CryptoTransfer` for signing requirements

Implement the below method in `CryptoPreTransactionHandlerImpl`
```
    @Override
    /** {@inheritDoc} */
    public TransactionMetadata preHandleCryptoTransfer(TransactionBody txn) {
        throw new NotImplementedException();
    }

### PR DESCRIPTION
**Pull request by georg-getz**
Signed-off-by: Georgi Georgiev <georgi.getz@outlook.com>

Adds implementation for the `preHandleCryptoTransfer` function.

Fixes #4228

### CODE REVIEW

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 1]
```diff
@@ -44,12 +44,12 @@ public CryptoPreTransactionHandlerImpl(@Nonnull final AccountStore accountStore)
 
     @Override
     /** {@inheritDoc} */
-    public TransactionMetadata preHandleCryptoCreate(final TransactionBody tx) {
-        final var op = tx.getCryptoCreateAccount();
+    public TransactionMetadata preHandleCryptoCreate(final TransactionBody txn) {
+        final var op = txn.getCryptoCreateAccount();
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:48]: Missing null check for txn

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 2]
```diff
@@ -44,12 +44,12 @@ public CryptoPreTransactionHandlerImpl(@Nonnull final AccountStore accountStore)
...
         final var key = asHederaKey(op.getKey());
         final var receiverSigReq = op.getReceiverSigRequired();
-        final var payer = tx.getTransactionID().getAccountID();
-        return createAccountSigningMetadata(tx, key, receiverSigReq, payer);
+        final var payer = txn.getTransactionID().getAccountID();
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:51]: Can txn.getTransactionID() be null?

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImplTest.java [hunk 3]
```diff
@@ -478,6 +503,21 @@ private TransactionBody cryptoDeleteAllowanceTransaction(final AccountID id) {
                 .build();
     }
 
+    private TransactionBody cryptoTransferTransaction(final AccountID id) {
+        if (id.equals(payer)) {
```
Comment (hendrikebbers) [inline hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImplTest.java:507]:
```suggestion
        if (Objects.equals(payer, id)) {
```

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java (hunk 1 above)
Comment (rbair23) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:48]: I also recommend adding @Nonnull annotations

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 4]
```diff
@@ -121,7 +121,8 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
     @Override
     /** {@inheritDoc} */
     public TransactionMetadata preHandleCryptoTransfer(TransactionBody txn) {
-        throw new NotImplementedException();
+        final var payer = txn.getTransactionID().getAccountID();
+        return new SigTransactionMetadata(accountStore, txn, payer);
     }
```
Comment (rbair23) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:126]: Should this method be somehow calling `createAccountSigningMetadata`? Can a crypto transfer work with accounts that have receiverSigRequired and we need to check that?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 5]
```diff
@@ -121,7 +121,8 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
     @Override
     /** {@inheritDoc} */
     public TransactionMetadata preHandleCryptoTransfer(TransactionBody txn) {
-        throw new NotImplementedException();
+        final var payer = txn.getTransactionID().getAccountID();
+        return new SigTransactionMetadata(accountStore, txn, payer);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:125]: We should be doing all the checks currently present at SigRequirements checks [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/sigs/order/SigRequirements.java#L674)

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Hi @georg-getz, as @Neeharika-Sompalli noted, each operation's `preHandler` must capture the exact signing requirements currently enforced by `SigRequirements` (for `CryptoTransfer`, [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/sigs/order/SigRequirements.java#L674)).

Because `CryptoTransfer` signing requirements involve looking up token metadata, this will be probably the most complicated handler to implement.

File: hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImplTest.java (hunk 3 above)
Comment (georg-getz) [reply hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImplTest.java:507]: What's wrong with using equals on the class itself? It is overridden in `AccountID`.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java (hunk 2 above)
Comment (georg-getz) [reply hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:51]: It cannot be, it explicitly checks for null and returns default instance if true.

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 6]
```diff
@@ -139,19 +151,96 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
...
+                    final var isUnapprovedDebit =
+                            accountAmount.getAmount() < 0 && !accountAmount.getIsApproval();
+                    if (isUnapprovedDebit) {
+                        meta.addNonPayerKey(accountAmount.getAccountID());
+                    } else {
+                        meta.addNonPayerKeyIfReceiverSigRequired(
+                                accountAmount.getAccountID(), INVALID_TRANSFER_ACCOUNT_ID);
+                    }
+                }
+            }
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:174]: It seems[ these checks](https://github.com/hashgraph/hedera-services/blob/9bf735718c899586a2839c252d11d98316b77567/hedera-node/hedera-mono-service/src/main/java/com/hedera/services/sigs/order/SigRequirements.java#L1294) to decide if the error status can be ignored in the account lookup from `SigRequirements` are missing ?

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 7]
```diff
@@ -139,19 +151,96 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
...
+
+            for (NftTransfer nftTransfer : transfers.getNftTransfersList()) {
+                if (!tokenMeta.failed() && nftTransfer.hasSenderAccountID()) {
+                    final var shouldAddSenderKey =
+                            !nftTransfer.getSenderAccountID().equals(payer)
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:179]: This check is already happening in `addNonPayerKey`

File: hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java [hunk 8]
```diff
@@ -139,19 +151,96 @@ public TransactionMetadata preHandleUpdateAccount(TransactionBody txn) {
...
+                                    && !nftTransfer.getIsApproval();
+                    if (shouldAddSenderKey) {
+                        meta.addNonPayerKey(nftTransfer.getSenderAccountID());
+                    }
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-mono-service/src/main/java/com/hedera/node/app/service/token/impl/CryptoPreTransactionHandlerImpl.java:183]: Seems like autoAccountCreation cases are not taken care ?

**Review by tinker-michaelj — CHANGES_REQUESTED** [review]
Thanks for tackling this one @georg-getz! 🙏 

The `CryptoTransfer` logic is complicated enough that it's hard to be sure a review will catch any edge-case deviation from current signing requirements.

I have created https://github.com/hashgraph/hedera-services/issues/4420 to refactor [`SigRequirementsTest`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/test/java/com/hedera/node/app/service/mono/sigs/order/SigRequirementsTest.java#L351) for use in testing the `PreTransactionHandler`'s and assigned myself; I will aim by EOD tomorrow to open a PR to your branch with the [`CryptoTransferScenarios`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-mono-service/src/testFixtures/java/com/hedera/test/factories/scenarios/CryptoTransferScenarios.java#L34) ported to further validate your new method.

This will also save us a lot of time in writing future unit tests.

**Comment by georg-getz** [comment]
Merged in #4567

**Pull request closed without merging**